main/dapps/server/asset/index.ts

```diff
@@ -1,96 +1,35 @@
-import cheerio from 'cheerio'
-import log from 'electron-log'
+import fs from 'fs'
+import path from 'path'
+import { app } from 'electron'
 
-import nebulaApi from '../../../nebula'
-import store from '../../../store'
 import getType from './getType'
 import { ServerResponse } from 'http'
 
-const nebula = nebulaApi()
-
 function error(res: ServerResponse, code: number, message: string) {
   res.writeHead(code || 404)
   res.end(message)
 }
 
-function getCid(namehash: string): string {
-  return store(`main.dapps`, namehash, `content`)
-}
+const dappCache = path.join(app.getPath('userData'), 'DappCache')
 
 export default {
-  stream: async (res: ServerResponse, namehash: string, path: string) => {
-    // Stream assets from IPFS back to the client
-    let found = false
-
-    const cid = getCid(namehash)
-
-    try {
-      for await (const chunk of nebula.ipfs.cat(`${cid}${path}`)) {
-        if (!found) {
-          res.setHeader('content-type', getType(path))
-          res.setHeader('Access-Control-Allow-Origin', '*')
-          res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,content-type')
-          res.writeHead(200)
-
-          found = true
-        }
-
-        res.write(chunk)
+  stream: async (res: ServerResponse, namehash: string, asset: string) => {
+    if (asset === '/') asset = '/index.html'
+    const assetPath = path.join(dappCache, namehash, asset)
+    if (fs.existsSync(assetPath)) {
+      try {
+        const stream = fs.createReadStream(assetPath)
+        res.setHeader('content-type', getType(asset))
+        res.setHeader('Access-Control-Allow-Origin', '*')
+        res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,content-type')
+        res.writeHead(200)
+        stream.pipe(res)
```

Comment:
@floating do we need to handle error events emitted from stream here or is that handled downstream?

Reply:
Ah yeah, we need to handle the error events for both streams, this and the `dapp.pipe` stream.